File: src/chain-observers/filters/transaction-id-filter.ts
import { BlockClassifier } from "../classifiers";
import type { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import type { TFilterEvaluationContext } from "../factories/data-evaluation-context";
import { FilterBase } from "./filter-base";

export class TransactionIdFilter extends FilterBase {
  public constructor(
    transactionIds: string[]
  ) {
    super();

    this.transactionIds = new Set(transactionIds);
  }

  private readonly transactionIds: Set<string>;

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      BlockClassifier
    ];
  }

  public async match(data: TFilterEvaluationContext): Promise<boolean> {
    const block = await data.get(BlockClassifier);

    for(const transactionId of this.transactionIds)
      if(block.transactionsPerId.has(transactionId))
        return true;

    return false;
  }
}

File: python/workerbee/chain_observers/filters/transaction_id_filter.py
"""TransactionIdFilter — matches transactions by their ID.

Mirrors src/chain-observers/filters/transaction-id-filter.ts.

Uses BlockClassifier.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._ordered_set import OrderedSet
from .filter_base import FilterBase

if TYPE_CHECKING:
    from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
    from ..factories.data_evaluation_context import DataEvaluationContext


class TransactionIdFilter(FilterBase):
    def __init__(self, transaction_ids: list[str]) -> None:
        super().__init__()
        self._transaction_ids: OrderedSet[str] = OrderedSet(transaction_ids)

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        from ..classifiers.block_classifier import BlockClassifier

        return [BlockClassifier]

    async def match(self, data: DataEvaluationContext) -> bool:
        from ..classifiers.block_classifier import BlockClassifier

        block = await data.get(BlockClassifier)

        return any(transaction_id in block["transactions_per_id"] for transaction_id in self._transaction_ids)
